src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -366,36 +361,54 @@ public async Task SaveChangesAsync(IEnumerable<UnspentOutputs> unspentOutputs, I
                     {
-                        // We'll need to restore the original outputs, so we clone it
-                        // and save it in rewind data.
-                        UnspentOutputs clone = unspent.Clone();
+                        // This can happen very rarely in case we fetch items form
+                        // disk and immedietly after the Evict method is called and removes the cached items.
 
-                        if (cacheItem.UnspentOutputs != null)
-                        {
-                            clone.Outputs = cacheItem.UnspentOutputs.Outputs.ToArray();
-                            rewindData.OutputsToRestore.Add(clone);
-                        }
-                        else
-                        {
-                            rewindData.TransactionsToRemove.Add(unspent.TransactionId);
-                        }
+                        this.logger.LogTrace("Outputs of transaction ID '{0}' are not found in cache, creating them.", unspent.TransactionId);
 
+                        FetchCoinsResponse result = await this.inner.FetchCoinsAsync(new[] {unspent.TransactionId}).ConfigureAwait(false);
+
+                        UnspentOutputs unspentOutput = result.UnspentOutputs[0];
+
+                        cacheItem = new CacheItem();
+                        cacheItem.ExistInInner = unspentOutput != null;
+                        cacheItem.IsDirty = false;
+
+                        cacheItem.UnspentOutputs = unspentOutput?.Clone();
+
+                        this.unspents.TryAdd(unspent.TransactionId, cacheItem);
+                    }
+                    else
+                    {
                         this.logger.LogTrace("Outputs of transaction ID '{0}' are in cache already, updating them.", unspent.TransactionId);
-                        if (cacheItem.UnspentOutputs != null) cacheItem.UnspentOutputs.Spend(unspent);
-                        else cacheItem.UnspentOutputs = unspent;
+                    }
+
+                    // If cacheItem.UnspentOutputs is null this means the trx was not stored in the disk,
+                    // that means the trx (and UTXO) is new and all the UTXO need to be stored in cache
+                    // otherwise we store to cache only the UTXO that have been spent.
+
+                    if (cacheItem.UnspentOutputs != null)
+                    {
+                        // To handle rewind we'll need to restore the original outputs,
+                        // so we clone it and save it in rewind data.
+                        UnspentOutputs clone = unspent.Clone();
+
+                        // We take the original items that are in cache and put them in rewind data.
+                        clone.Outputs = cacheItem.UnspentOutputs.Outputs.ToArray();
+                        rewindData.OutputsToRestore.Add(clone);
+
+                        // Now modify the cached items with the mutated data.
```

Comment:
What do you mean by mutated data ? Is the the code below setting the `unspent` item as **spent** ?

Reply:
Yes correct its marking as spent the UTXO that have been marked as spent after passing in the rules engine.